sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/SsrfInjection.kt

```diff
@@ -0,0 +1,157 @@
+ */
+@Suppress("unused_parameter", "unused")
+object SsrfInjection {
+
+    /**
+     * Honeypot host name targeting an invalid address.
+     * RFC 2606 defines such names at https://www.rfc-editor.org/rfc/rfc2606#section-2
+     */
+    private const val HONEYPOT_HOST: String = "jazzer.invalid"
+
+    // [java.nio.channels.SocketChannel] is used in many JDK classes to open
+    // (non-blocking) network connections. E.g. [java.net.http.HttpClient] uses
+    // it internally.
+    //
+    // The actual connection is established in the abstract "connect" method.
+    // Hooking that also hooks all concrete implementations, from which
+    // only one exists in [sun.nio.ch.SocketChannelImpl]. "connect" is only
+    // called in [java.nio.channels.SocketChannel] itself and the two
+    // mentioned classes.
+    @MethodHook(
+        type = HookType.BEFORE,
+        targetClassName = "java.nio.channels.SocketChannel",
+        targetMethod = "connect",
+        additionalClassesToHook = [
+            "sun.nio.ch.SocketAdaptor",
+            "jdk.internal.net.http.PlainHttpConnection",
+        ],
+    )
+    @JvmStatic
+    fun checkSsrfHttpConnection(method: MethodHandle, thisObject: Any?, arguments: Array<Any>, hookId: Int) {
+        checkSsrf(arguments, hookId)
+    }
+
+    // [java.net.Socket] is used in many JDK classes to open network connections.
+    // Internally it delegates to [java.net.SocketImpl], hence, for most situations
+    // it's sufficient to hook [java.net.Socket] itself.
+    //
+    // As [java.net.SocketImpl] is an abstract class, so all concrete implementations
+    // get hooked. Internal classes invoking connect on those implementations need
+    // to be specified explicitly, though.
+    //
+    // Internal classes "using" [java.net.SocketImpl.connect]:
+    // - java.net.Socket (hook required)
+    // - java.net.AbstractPlainSocketImpl (no direct usage)
+    // - java.net.PlainSocketImpl (no direct usage)
+    // - java.net.HttpConnectSocketImpl (only used in Socket)
+    // - java.net.SocksSocketImpl (used in Socket, but also invoking super.connect directly, hook required)
+    // - java.net.ServerSocket (security check, no hook required)
+    @MethodHook(
+        type = HookType.BEFORE,
+        targetClassName = "java.net.SocketImpl",
+        targetMethod = "connect",
+        additionalClassesToHook = [
+            "sun.net.NetworkClient",
+            "java.net.Socket",
+            "java.net.SocksSocketImpl",
+            "java.net.NetworkInterface",
+        ],
+    )
+    @JvmStatic
+    fun checkSsrfSocket(method: MethodHandle, thisObject: Any?, arguments: Array<Any>, hookId: Int) {
+        checkSsrf(arguments, hookId)
+    }
+
+    @JvmStatic
+    fun checkSsrf(arguments: Array<Any>, hookId: Int) {
+        if (arguments.isEmpty()) {
+            return
+        }
+
+        val host = when (val arg = arguments[0]) {
```

Comment:
Couldn't we collapse this block into:

```java
val host = when (val arg = arguments[0]) {
    is String -> try {
        URI.create(arg).host
    } catch (e: Exception) {
        null
    }
    is InetAddress, is InetSocketAddress -> arg.hostName
    else -> null
} ?:return
```

The try-catch block can be expensive if an invalid URL is provided. Instead of catching the exception, we can use the `URI.create` which AFAIK should throw an exception if the provided URL is invalid.

Also, combining `InetAddress` and `InetSocketAddress` seems reasonable

Reply:
Can you elaborate how switching from `URL` to `URI` would improve the performance? The try-catch-block would still be needed, wouldn't it?

Indeed, combining `InetAddress` and `InetSocketAddress` would be nice, but the compiler can not derive a common type, beside `Object`, and so `hostName` wouldn't be accessible anymore.

Reply:
Ah that makes sense. 

Regarding the first point: I was sure `URI.create()` allows for unchecked exceptions which would be cheaper (as we will run into this exception a lot) and would allow us to get rid of the try-catch block. Upon further inspection of the OpenJDK code I was wrong. The only difference here is that using URI could cover different aspects, as every URL is an URI, but not vice versa. In the end as we're specifically testing for URLs here using `URL()` should be fine.

`URI` uses [parse](https://github.com/openjdk/jdk/blob/52d30087734ad95761078793da6e207797558e2b/src/java.base/share/classes/java/net/URI.java#L3152)
`URL` ends up [here](https://github.com/openjdk/jdk/blob/52d30087734ad95761078793da6e207797558e2b/src/java.base/share/classes/java/net/URL.java#L490) 

If we really  wanted, we could check which one is faster to fail, as we'd be triggering the bad path over and over again..